### ISSUE TITLE
Accessing unused encoding slots in Python throws TypeError, instead of returning False for "font[slot].isWorthOutputting()". 

### ISSUE THREAD

**Issue by HinTak** [fontforge/fontforge#3107]
While re-writing a script in the legacy scripting interface to the python scripting interface 
( https://github.com/fontforge/fontforge/issues/3085#issuecomment-312740591 ), I came across this wierdness of the python interface: access an encoding slot that has not been used, results in a TypeError,
instead of a glyph which is false for "font[slot].isWorthOutputting()". This is not or poorly documented from in the slot accessing documentation, or the "isWorthOutputting()" documentation.

Also, this makes testing for unused slots (i.e. Not WorthOutputing()) a bit bizzare, as throwing an TypeError is the expected behavior while Not Throwing is the surprise and Not Throwing is therefore a error condition.

**Comment by khaledhosny** [comment]
From the documentation:

> isWorthOutputting 	() 	Returns whether the glyph is worth outputting into a font file. Basically a glyph is worth outputting if it contains any contours, or references or has had its width set.

I don’t think it does what you think it is doing.

**Comment by HinTak** [comment]
My (intended) usage of the "{Is}WorthOutputting()" is simply to confirm that I am copying from a source used slot into a destination unused slot.This works with the legacy scripting interface. But in the python code, the expected outcome (unused slot) throws a TypeError, while the surprise outcome, for which the user needs to be alerted, is when it DOES NOT throw an TypeError exception.
So I needed to both suppress the TypeError, and also deliberately throw another Error (currently I just print a message and continue) when fontforge does not throw a TypeError.

Retrieving unused slots - a glyph position that has not been drawn on - gives a TypeError is certainly surprising. It should just be a largely empty glyph, rather than throwing a TypeError.

This is also very confusing when the python routine shares almost the same name (except the "Is" at the beginning) with the legacy scripting routine but behaves substantially differently.

**Comment by khaledhosny** [comment]
The `TypeError` is if coming from trying to index the font with a non-existing glyph; i.e. `font[slot]`, the index should be a glyph name or a glyph encoding (font specific, a fontforge thing not to be confused with the unicode value). `isWorthOutputting()` method is a member of the glyph class, you just can’t use it without a glyph.

**Comment by HinTak** [comment]
That still does not help. I know what the current behavior is. However, let me repeat, that the goal is to test for unused destination (to paste into). Unfortunately, the desired outcome is an error condition in fontforge, and the no-error situation is the undesirable outcome that
one must warn the user. That means the client program needs to suppress an fontforge exception, and also intentionally throw an error WHENEVER fontforge does NOT give an error.

And, also, testing for unused destination to paste into, works under the legacy scripting interface.

**Comment by frank-trampe** [comment]
@HinTak, as @khaledhosny has explained, one of the side effects of making the glyphs available in a native Python dictionary is that you cannot access non-existent members and that there is nothing that FontForge can do about it. It seems easy enough to me for you to use a wrapper function in this case.

```
def worthOutputting(PySF, glyph):
  if glyph in PySF:
    return PySF[glyph].isWorthOutputting()
  return 0
```

**Comment by HinTak** [comment]
@frank-trampe : yes, your suggestion sounds good. In fact I think the font object does not have a worthoutputting() method - why can't such a wrapper added, to mean font.selection.isworthoutputting() ? That also matches how the legacy scripting interface closer too. Btw, I am sure you mean `return False` instead of zero.

**Comment by HinTak** [comment]
I went ahead and added it :
https://github.com/fontforge/fontforge/pull/3109

**Comment by HinTak** [comment]
BTW, feta is a font for music symbols, so reencode to ucs4 means no encoding slots are used; but you can still select by the symbol names.

**Comment by JoesCat** [comment]
> How come the whole python extension is written in C, AFAIC? It would ne nice to do it in hybrids - i.e. some methods in C and some python....

The initial is_Ligatures introduced hard dependencies on python that brought all sorts of grief when some distros were at 2.7 while mainline distros like Debian were still at python 2.6. The quick fix to quiet the issues was to kill the python dependency, but we had to eventually bring is_Ligatures inline so that it could also be updated easily (now has no python dependency).
I'm looking at eventually replacing Unicode/charset perl dependency too (plus it misses a couple files).

When you put too many dependencies, it becomes harder to adopt something new, and FontForge is a bit of a monster that has taken a long time to move up from 20120731 for some distros (like pulling teeth), and is in fact still the choice for some distros, some by choice, but I suspect some also due to the large hurdle to jump.
In terms of dependencies, I'm in favour of reducing rather than increasing them.

**Comment by HinTak** [comment]
Ghostscript is part postscript and part C ( I am one of them) and R is part R and part C also ( I am not one of them but had written quite extensively in R) . There is no reason why the python extension to fontforge cannot be part python and part C. That would allow all those mediocre python people who cannot write C code easier to contribute, just my 2 cents. 

Or, improve the native scripting interface . It is certainly weird that nobody is working on the native scripting interface, and the python interface is all written in C which sets a barrier to mediocre people to contribute.

**Comment by JoesCat** [comment]
There are some very bright people running native script, and likewise, I'm not to beating myself over the head running a secure distro like openBSD when I can run something that has minimum fuss like say, Mageia, or Mandrake-similar.
In terms of nobody working on native scripting, fontanvil was created precisely because of a prior proposal to remove nativescripting, and #3026 is the latest PR that adds a function.

I would choose "time constrained" versus "mediocre", but you'll understand this when you're older  ;-P
If given insufficient time for a task, you can guarantee failure, no matter how good you are, and fact is, a lot of people have to focus their 24x7 in directions best suited to them with time on hand. ;-)
